### PR TITLE
PLANNER-1825 Add constraint provider for TSP

### PR DESCRIPTION
**Pull request by triceo**
Example verified with assertion score director.
CS-D performance on par with DRL.

### CODE REVIEW

File: optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/app/TspApp.java [hunk 1]
```diff
@@ -55,7 +55,7 @@ protected TspPanel createSolutionPanel() {
 
     @Override
     public SolutionFileIO<TspSolution> createSolutionFileIO() {
-        return new XStreamSolutionFileIO<>(TspSolution.class);
+        return new TspXStreamSolutionFileIO();
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/app/TspApp.java]: Why extract that in a separate file? The more files in an example, the harder it is to understand for our users.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/optional/score/TspConstraintProvider.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        return new Constraint[]{
+                distanceToPreviousStandstill(constraintFactory),
+                distanceFromLastVisitToDomicile(constraintFactory)
+        };
+    }
+
+    private Constraint distanceToPreviousStandstill(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Visit.class)
+                .penalizeLong("Distance to previous standstill",
+                        SimpleLongScore.of(1),
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/optional/score/TspConstraintProvider.java]: Use SimpleLongScore.ONE

File: optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/app/TspApp.java (hunk 1 above)
Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/app/TspApp.java]: I thought you were going to ask. :-) I have a harness that allows me to run all examples in a single benchmark - essentially an abstraction to easily perform score director benchmarks over all examples and all score directors, without having a benchmark template of a million lines.

The downside of that is that it relies pretty heavily on generics. And I had to do this to standardize it with the other examples. (This was the only one that didn't have its own solution file IO.)

I will submit the harness as a PR later and we can discuss it.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/optional/score/TspConstraintProvider.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                        Visit::getDistanceFromPreviousStandstill);
+    }
+
+    private Constraint distanceFromLastVisitToDomicile(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Visit.class)
+                .ifNotExists(Visit.class, Joiners.equal(visit -> visit, Visit::getPreviousStandstill))
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/optional/score/TspConstraintProvider.java]: I guess "visit -> visit" is better than "Function.identity()"?

Should we experiment with thisEqual() or equalToThat() ? I guess that's overkill...

File: optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/app/TspApp.java (hunk 1 above)
Comment (ge0ffrey) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/app/TspApp.java]: If all other examples do it, there is very little point indeed in this example not doing it to be easier to understand. +1 for this change (consistency)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/optional/score/TspConstraintProvider.java (hunk 3 above)
Comment (ge0ffrey) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/optional/score/TspConstraintProvider.java]: (just a thought, no changes needed)

Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/tsp/optional/score/TspConstraintProvider.java]: In this particular case, `Function.identity()` didn't compile.`identity()` returns a `Visit` while `getPreviousStandstill()` returns a `Standstill`. The compiler didn't like that.

**Pull request merged**